File: app/models/Project.py
from app.services.mouli_service import MouliService


class Project:
    _id: str = None

    code_acti: str
    student_id: str
    title: str

    date_start: str
    date_end: str

    code_module: str
    title_module: str

    fetch_date: str

    mouli_project_code: str | None = None
    slug: str | None = None


    def __init__(self, mongo_data=None):
        if mongo_data is None:
            return
        self._id = mongo_data["_id"]
        self.code_acti = mongo_data["code_acti"]
        self.student_id = mongo_data["student_id"]
        self.date_start = mongo_data["date_start"]
        self.date_end = mongo_data["date_end"]
        self.code_module = mongo_data["code_module"]
        self.title = mongo_data["title"]
        self.title_module = mongo_data["title_module"]
        self.fetch_date = mongo_data["fetch_date"]
        self.slug = mongo_data.get("slug", None)

    def to_dict(self):
        return {
            "_id": self._id,
            "code_acti": self.code_acti,
            "student_id": self.student_id,
            "date_start": self.date_start,
            "date_end": self.date_end,
            "code_module": self.code_module,
            "title_module": self.title_module,
            "title": self.title,
            "fetch_date": self.fetch_date,
            "slug": self.slug
        }

    def to_api(self):

        latest_mouli = MouliService.get_latest_of_project(self.slug, self.student_id)

        return {
            "code_acti": self.code_acti,
            "date_start": self.date_start,
            "date_end": self.date_end,
            "code_module": self.code_module,
            "title_module": self.title_module,
            "title": self.title,
            "fetch_date": self.fetch_date,
            "slug": self.slug,
            "mouli": {
                "test_id": latest_mouli.test_id,
                "score": latest_mouli.score,
                "date": latest_mouli.test_date
            } if latest_mouli else None
        }

    @property
    def mongo_id(self):
        return self._id

File: web/src/models/Project.ts
export class EpiProject {
    acti_code: string;
    project_name: string;
    project_slug: string;
    module_code: string;
    module_name: string;
    start_date: Date;
    end_date: Date;
    mouli_seen: boolean;
    mouli: {
        test_id: number;
        score: number;
        date: Date;
        is_warning: boolean;
    } | null;

    constructor(data: any) {
        this.acti_code = data.code_acti;
        this.project_name = data.title;
        this.project_slug = data.slug;
        this.module_code = data.code_module;
        this.module_name = data.title_module;
        this.start_date = data.date_start;
        this.end_date = data.date_end;
        this.mouli_seen = data.mouli_seen;
        if (data.mouli) {
            this.mouli = {
                test_id: data.mouli.test_id,
                score: data.mouli.score,
                date: data.mouli.date,
                is_warning: data.mouli.is_warning,
            }
        } else {
            this.mouli = null;
        }
    }


}